provisioning/main.c

```diff
@@ -0,0 +1,321 @@
+        return;
+    }
+
+    output = calloc(1, OUTPUT_LEN);
+    if (!output) {
+        puts("Out of memory");
+        return;
+    }
+
+    mbedtls_pk_init(&pk);
+    ret = mbedtls_pk_setup_opaque(&pk, handle);
+    if (ret != 0)
+    {
+        printf("Failed to setup PK with ret=%d\n", ret);
+        goto done;
+    }
+
+    ret = mbedtls_pk_write_pubkey_pem(&pk, output, OUTPUT_LEN);
+    if (ret != 0) {
+        printf("Failed to print pubkey with ret=%d\n", ret);
+        goto done;
+    }
+
+    printf("%s\n", output);
+
+done:
+    mbedtls_pk_free(&pk);
+    free(output);
+}
+
+void generate_csr(psa_key_id_t key_id)
+{
+    int ret;
+    unsigned char *output;
+    enum { OUTPUT_LEN = 2048 };
+    psa_status_t status;
+    psa_key_handle_t handle;
+    mbedtls_pk_context pk;
+    mbedtls_x509write_csr req;
+
+    /* Initialize Mbed TLS structures. */
+    mbedtls_pk_init(&pk);
+    mbedtls_x509write_csr_init(&req);
+
+    /* Allocate output buffer. */
+    output = calloc(1, OUTPUT_LEN);
+    if (!output)
+    {
+        puts("Out of memory");
+        return;
+    }
+
+    /* Open the specified key. */
+    status = psa_open_key(key_id, &handle);
+    if (status != PSA_SUCCESS)
+    {
+        printf("Failed to open key %lu with status=%ld\n", key_id, status);
+        goto done;
+    }
+
+    ret = mbedtls_pk_setup_opaque(&pk, handle);
+    if (ret != 0)
+    {
+        printf("Failed to setup PK with ret=%d\n", ret);
+        goto done;
+    }
+
+    mbedtls_x509write_csr_set_md_alg(&req, MBEDTLS_MD_SHA256);
+
+    ret = mbedtls_x509write_csr_set_subject_name(
+        &req, "CN=Device,O=Mbed TLS,OU=client,C=UK");
+    if (ret != 0)
+    {
+        printf("Failed to set subject name with ret=%d\n", ret);
+        goto done;
+    }
+
+    mbedtls_x509write_csr_set_key_usage(
+        &req, MBEDTLS_X509_KU_DIGITAL_SIGNATURE);
+
+    mbedtls_x509write_csr_set_ns_cert_type(
+        &req, MBEDTLS_X509_NS_CERT_TYPE_SSL_CLIENT);
+
+    mbedtls_x509write_csr_set_key(&req, &pk);
+
+    ret = mbedtls_x509write_csr_pem(&req, output, OUTPUT_LEN,
+        psa_rng_for_mbedtls, NULL);
+    if (ret != 0)
+    {
+        printf("Failed to make CSR with ret=%d\n", ret);
+        goto done;
+    }
+
+    printf("\tKey ID %lu:\n", key_id);
+    printf("%s\n", output);
+
+done:
+    free(output);
+    mbedtls_x509write_csr_free(&req);
+    mbedtls_pk_free(&pk);
+}
+
+/* The secure element factory put a device private key pair (not attestation
+ * key) into a slot in the secure element. We need to tell Mbed Crypto that
+ * this key pair exists so that it can be used. */
+
+/* Put trusted root CA certificate into persistent storage */
+/* When and how should PSA use this? TLS wants for authenticating the server. */
+/* Application will have this, not provisioning. */
+
+int main(void)
+{
+    psa_status_t status;
+    printf("Provisioning device...\n");
+
+    printf("\tErasing device... ");
```

Comment:
This isn't quite erasing the device — it's not like it's changing the code image. Please clarify what gets erased, and also that this may involve a reboot.